modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/requestcontroller/filter/NuxeoRequestControllerFilter.java

```diff
@@ -295,4 +319,34 @@ protected void addCacheHeaders(HttpServletResponse response, RequestFilterConfig
         }
     }
 
+    protected void handleRequest(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        if (request.getMethod().equals("GET") || request.getMethod().equals("POST")) {
+            String key = request.getHeader(IDEMPOTENCY_KEY);
+            if (key != null) {
+                KeyValueService kvs = Framework.getService(KeyValueService.class);
+                KeyValueStore store = kvs.getKeyValueStore(idempotencyStoreName);
+                String value = store.getString(key);
+                if (value == null) {
+                    // new request
+                    store.put(key, IDEMPOTENCY_INPROGRESS_MARKER, idempotencyTTLSeconds);
+                    CopyingHttpServletResponse wrapper = new CopyingHttpServletResponse(response);
+                    chain.doFilter(request, wrapper);
+                    store.put(key, wrapper.getCopy(), idempotencyTTLSeconds);
+                } else if (IDEMPOTENCY_INPROGRESS_MARKER.equals(value)) {
+                    // request already in progress -> conflict (?) any content?
+                    response.sendError(HttpServletResponse.SC_CONFLICT);
+                } else {
+                    // request already done: return stored result
+                    // XXX get back status code?
```

Comment:
https://github.com/paypal/api-standards/blob/master/patterns.md#idempotency explains that should return HTTP 200 + the result.

When it is a new request, it should return HTTP 201  + the result.

I am not familiar with the code here, does that return such HTTP codes?

Reply:
I went with conflict because that's what's documented here https://tools.ietf.org/html/draft-idempotency-header-01.
As for 201, this filter cannot know this is for creation so no sure 201 applies here, so the code is left unchanged compared to a request that would go through without idempotency.

Reply:
Conflict seems the good error to throw.

200 or 201, I would say this is fine as-is. As soon as this is not an error code :)

Reply:
Well it will be an error code if there's indeed an error while processing the request.
That's also why we may want the code to be stored too the first time that the request is processed.

Reply:
Hmm how would you make the difference between an HTTP 500 error and a "valid" error like a 404? The 500 error should not be cached as another call may unlock the real result. And a 404 now may not be revelant 2 minutes later. Same for 401 and 403 errors for example.

Reply:
Ok so maybe the code should be checked before storing the result, and key removed from the store in this case?

Reply:
I changed the processing so that in case of error, store content is nullified for this key